File: src/main/java/de/twometer/neko/res/TextureLoader.java
package de.twometer.neko.res;

import de.twometer.neko.gl.Cubemap;
import de.twometer.neko.gl.Texture;
import de.twometer.neko.util.CrashHandler;
import de.twometer.neko.util.Log;

import java.awt.image.BufferedImage;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.Arrays;

import static org.lwjgl.opengl.GL11.*;
import static org.lwjgl.opengl.GL11.GL_TEXTURE_2D;
import static org.lwjgl.opengl.GL12.GL_CLAMP_TO_EDGE;
import static org.lwjgl.opengl.GL12.GL_TEXTURE_WRAP_R;
import static org.lwjgl.opengl.GL13.GL_TEXTURE_CUBE_MAP;
import static org.lwjgl.opengl.GL13.GL_TEXTURE_CUBE_MAP_POSITIVE_X;
import static org.lwjgl.opengl.GL30.glGenerateMipmap;

public class TextureLoader {

    /**
     * Load a cubemap comprised of 6 textures from disk
     *
     * @param textures Textures in order POSX (right), NEGX (left), POSY (top), NEGY (bottom), POSZ (back), NEGZ (front)
     * @return The cubemap texture object
     */
    public static Cubemap loadCubemap(String... textures) {
        Log.d("Loading cubemap " + Arrays.toString(textures));

        try {
            int textureId = glGenTextures();
            glBindTexture(GL_TEXTURE_CUBE_MAP, textureId);

            for (int i = 0; i < textures.length; i++) {
                var image = ResourceLoader.loadImage(AssetPaths.TEXTURE_PATH + textures[i]);
                var pixels = ResourceLoader.loadPixels(image);
                glTexImage2D(GL_TEXTURE_CUBE_MAP_POSITIVE_X + i, 0, GL_RGB, image.getWidth(), image.getHeight(), 0, GL_RGBA, GL_UNSIGNED_BYTE, pixels);
            }

            glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MAG_FILTER, GL_LINEAR_MIPMAP_LINEAR);
            glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
            glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
            glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
            glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_R, GL_CLAMP_TO_EDGE);
            glGenerateMipmap(GL_TEXTURE_CUBE_MAP);

            return new Cubemap(textureId);
        } catch (IOException e) {
            CrashHandler.fatal(e);
            return null;
        }
    }

    public static Texture loadTexture(String name) {
        Log.d("Loading texture " + name);
        var path = AssetPaths.TEXTURE_PATH + name;
        try {
            BufferedImage image = ResourceLoader.loadImage(path);
            ByteBuffer buffer = ResourceLoader.loadPixels(image);

            int textureId = glGenTextures();
            glBindTexture(GL_TEXTURE_2D, textureId);
            glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, image.getWidth(), image.getHeight(), 0, GL_RGBA, GL_UNSIGNED_BYTE, buffer);
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR_MIPMAP_LINEAR);
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
            glGenerateMipmap(GL_TEXTURE_2D);

            glBindTexture(GL_TEXTURE_2D, 0);

            return new Texture(textureId, image.getWidth(), image.getHeight());
        } catch (IOException e) {
            CrashHandler.fatal(e);
            return null;
        }
    }



}

File: neko-engine/src/main/kotlin/de/twometer/neko/res/TextureLoader.kt
package de.twometer.neko.res

import de.twometer.neko.render.Texture2d
import mu.KotlinLogging
import org.lwjgl.opengl.GL11.*
import org.lwjgl.opengl.GL30.glGenerateMipmap
import java.nio.ByteBuffer

private val logger = KotlinLogging.logger {}

object TextureLoader {

    fun load(path: String): Texture2d {
        val filename = AssetManager.resolve(path, AssetType.Textures).absolutePath
        logger.debug { "Loading texture $path" }

        val image = ImageLoader.load(filename)
        val texture = load(image.pixels, image.width, image.height)
        return texture.also { image.destroy() }
    }

    fun load(pixels: ByteBuffer, width: Int, height: Int, mipmap: Boolean = true): Texture2d {
        val textureId = glGenTextures()
        glBindTexture(GL_TEXTURE_2D, textureId)

        glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, pixels)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, if (mipmap) GL_LINEAR_MIPMAP_LINEAR else GL_LINEAR)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT)
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT)

        if (mipmap)
            glGenerateMipmap(GL_TEXTURE_2D)

        glBindTexture(GL_TEXTURE_2D, 0)

        return Texture2d(textureId, width, height)
    }

}